File: tiq-e/linked-list/5/main.py
from time import perf_counter


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        previous = None
        while slow:
            nextnode = slow.next
            slow.next = previous
            previous = slow
            slow = nextnode

        while previous:
            if previous.val != head.val:
                return False
            previous = previous.next
            head = head.next
        return True


head = ListNode(1, None)

start = perf_counter()
result = Solution().isPalindrome(head)
end = perf_counter()
print("Duration: " + str(end - start))
print("Result: " + str(result))

File: tiq-e/linked-list/5/main.go
package main

import (
	"fmt"
	"time"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	head, slow, fast := nil, head, head
	for fast != nil && fast.Next != nil {
		head, slow.Next, slow, fast = slow, head, slow.Next, fast.Next.Next
	}
	if fast != nil {
		slow = slow.Next
	}
	for ; head != nil && head.Val == slow.Val; head, slow = head.Next, slow.Next {
	}
	return head == nil
}

func main() {
	var head = ListNode{Val: 1, Next: nil}

	var start = time.Now()
	var result = isPalindrome(&head)
	var end = time.Since(start)
	fmt.Println("Duration: ", end)
	fmt.Println("Result: ", result)
}
